app/controllers/application_controller.rb

```diff
@@ -5,6 +5,143 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :sessions_secret, "secret"
   end
 
-end
\ No newline at end of file
+# Users Actions
+  get '/' do
+    erb :index
+  end
+
+  get '/login' do
+    if Helpers.is_logged_in?(session)
+      redirect '/tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  get '/signup' do
+    if Helpers.is_logged_in?(session)
+
+      redirect '/tweets'
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    if !(params.any? {|key, userdata| userdata.empty?})
+      @user = User.create(params)
+      session[:id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/signup'
+    end
+  end
+
+  get '/logout' do
+    if Helpers.is_logged_in?(session)
+      session.clear
+      redirect '/login'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(username: params[:username])
+    session[:id] = user.id
+    redirect '/tweets'
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+#tweets actions
+  get '/tweets' do
```

Comment:
This controller is overloaded. Consider separating your concerns by splitting some of these actions into separate controllers (i.e. a users controller and a tweets controller).